File: gremlin-go/driver/util/profiling_application.go
/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/

package main

import (
	"flag"
	"fmt"
	"math"
	"math/rand"
	"os"
	"sync"
	"sync/atomic"
	"time"

	gremlingo "github.com/apache/tinkerpop/gremlin-go/v4/driver"
)

type scriptChooser struct {
	mu       sync.Mutex
	rng      *rand.Rand
	exercise bool
	script   string
}

func newScriptChooser(exercise bool, script string) *scriptChooser {
	return &scriptChooser{
		rng:      rand.New(rand.NewSource(0)),
		exercise: exercise,
		script:   script,
	}
}

func (sc *scriptChooser) next() string {
	if sc.exercise {
		sc.mu.Lock()
		idx := sc.rng.Intn(len(scripts) - 1)
		sc.mu.Unlock()
		return scripts[idx]
	}
	return sc.script
}

var scripts = []string{
	`g.V()`,
	`g.V(1).out('knows')`,
	`g.V(1).out('knows').has('name','josh')`,
	`g.V(1).as("a").out("knows").as("b").select("a", "b")`,
	`g.V(1).as("a").out("knows").as("b").select("a", "b").by("name")`,
	`g.V().hasLabel("person").as("p").map(__.bothE().label().groupCount()).as("r").select("p", "r")`,
	`g.V().choose(__.outE().count().is(0L), __.as("a"), __.as("b")).choose(__.select("a"), __.select("a"), __.select("b"))`,
	`g.V().group("a").by(T.label).by(outE().values("weight").sum()).cap("a")`,
	`g.V().repeat(__.union(__.out("knows").group("a").by("age"), __.out("created").group("b").by("name").by(count())).group("a").by("name")).times(2).cap("a", "b")`,
	`g.V().match(as("a").out("knows").as("b"),as("b").out("created").as("c"),as("c").in("created").as("d"),as("d").where(P.neq("a")).where(P.neq("b"))).select("a","b","c","d")`,
}

func main() {
	testType := flag.String("test-type", "throughput", "Test type: throughput or latency")
	host := flag.String("host", "localhost", "Gremlin server host")
	port := flag.Int("port", 8182, "Gremlin server port")
	parallelism := flag.Int("parallelism", 16, "Number of parallel goroutines")
	warmups := flag.Int("warmups", 5, "Number of warmup iterations")
	executions := flag.Int("executions", 10, "Number of test iterations")
	requests := flag.Int("requests", 10000, "Number of requests per iteration")
	script := flag.String("script", "g.inject(1)", "Gremlin script to execute")
	exercise := flag.Bool("exercise", false, "Exercise mode with Modern graph scripts")
	poolSize := flag.Int("pool-size", 8, "Connection pool size")
	tooSlowThreshold := flag.Int("too-slow-threshold", 125, "Too slow threshold in ms")
	timeout := flag.Int("timeout", 1200000, "Timeout in ms")
	minExpectedRps := flag.Int("min-expected-rps", 1000, "Minimum expected requests per second")
	pauseBetweenRuns := flag.Int("pause-between-runs", 1000, "Pause between runs in ms")
	store := flag.String("store", "", "TSV output file path")
	noExit := flag.Bool("no-exit", false, "Do not exit after completion")

	flag.Parse()

	url := fmt.Sprintf("http://%s:%d/gremlin", *host, *port)

	chooser := newScriptChooser(*exercise, *script)

	switch *testType {
	case "throughput":
		runThroughputTest(url, *parallelism, *warmups, *executions, *requests,
			*exercise, *poolSize, *tooSlowThreshold, *timeout, *minExpectedRps,
			*pauseBetweenRuns, *store, chooser)
	case "latency":
		runLatencyTest(url, *warmups, *executions, *exercise, *poolSize,
			*timeout, *pauseBetweenRuns, chooser)
	default:
		fmt.Fprintf(os.Stderr, "Unknown test type: %s\n", *testType)
		os.Exit(1)
	}

	if *noExit {
		select {}
	}
}

func createClient(url string, poolSize int) (*gremlingo.Client, error) {
	return gremlingo.NewClient(url, func(settings *gremlingo.ClientSettings) {
		settings.MaximumConcurrentConnections = poolSize
		settings.LogVerbosity = gremlingo.Error
	})
}

func initializeGraph(url string, poolSize int) {
	client, err := createClient(url, poolSize)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error creating client for initialization: %v\n", err)
		os.Exit(1)
	}
	defer client.Close()

	rs, err := client.Submit("graph.clear()")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error clearing graph: %v\n", err)
		os.Exit(1)
	}
	_, _ = rs.All()

	rs, err = client.Submit("TinkerFactory.generateModern(graph)")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error generating modern graph: %v\n", err)
		os.Exit(1)
	}
	_, _ = rs.All()
}

func runThroughputTest(url string, parallelism, warmups, executions, requests int,
	exercise bool, poolSize, tooSlowThreshold, timeout, minExpectedRps,
	pauseBetweenRuns int, store string, chooser *scriptChooser) {

	fmt.Println("-----------------------THROUGHPUT TEST SELECTED--------------------")

	if exercise {
		fmt.Println("--------------------------INITIALIZATION--------------------------")
		initializeGraph(url, poolSize)
	}

	fmt.Println("---------------------------WARMUP CYCLE---------------------------")
	warmupRequests := 1000
	var totalWarmupRps float64
	thresholdDuration := time.Duration(tooSlowThreshold) * time.Millisecond

	for i := 0; i < warmups; i++ {
		client, err := createClient(url, poolSize)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error creating client for warmup: %v\n", err)
			os.Exit(1)
		}

		start := time.Now()
		var wg sync.WaitGroup
		sem := make(chan struct{}, parallelism)

		for r := 0; r < warmupRequests; r++ {
			wg.Add(1)
			sem <- struct{}{}
			go func() {
				defer wg.Done()
				defer func() { <-sem }()
				s := chooser.next()
				rs, err := client.Submit(s)
				if err == nil {
					_, _ = rs.All()
				}
			}()
		}
		wg.Wait()

		elapsed := time.Since(start).Seconds()
		rps := float64(warmupRequests) / elapsed
		totalWarmupRps += rps
		fmt.Printf("[warmup-%d] requests: %d | time(s): %.3f    | req/sec: %d\n",
			i+1, warmupRequests, elapsed, int(math.Round(rps)))

		client.Close()
		time.Sleep(time.Duration(pauseBetweenRuns) * time.Millisecond)
	}

	if warmups > 0 {
		avgWarmupRps := totalWarmupRps / float64(warmups)
		if avgWarmupRps < float64(minExpectedRps) && !exercise {
			fmt.Printf("avg req/sec during warmup (%.0f) is below minimum expected (%d), skipping test cycles\n",
				avgWarmupRps, minExpectedRps)
			return
		}
	}

	fmt.Println("----------------------------TEST CYCLE----------------------------")
	var totalRps float64
	var completedExecs int
	testStart := time.Now()

	for i := 0; i < executions; i++ {
		if time.Since(testStart).Milliseconds() > int64(timeout) {
			fmt.Printf("Timeout reached, skipping remaining test cycles\n")
			break
		}

		client, err := createClient(url, poolSize)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error creating client for test: %v\n", err)
			os.Exit(1)
		}

		var tooSlow int64
		var errors int64
		start := time.Now()
		var wg sync.WaitGroup
		sem := make(chan struct{}, parallelism)

		for r := 0; r < requests; r++ {
			wg.Add(1)
			sem <- struct{}{}
			go func() {
				defer wg.Done()
				defer func() { <-sem }()
				reqStart := time.Now()
				s := chooser.next()
				rs, err := client.Submit(s)
				if err != nil {
					atomic.AddInt64(&errors, 1)
				} else {
					_, err = rs.All()
					if err != nil {
						atomic.AddInt64(&errors, 1)
					}
				}
				if time.Since(reqStart) > thresholdDuration {
					atomic.AddInt64(&tooSlow, 1)
				}
			}()
		}
		wg.Wait()

		elapsed := time.Since(start).Seconds()
		rps := float64(requests) / elapsed
		totalRps += rps
		completedExecs++

		if exercise {
			fmt.Printf("[test-%d]   requests: %d | time(s): %.3f    | req/sec: %d   | too slow: N/A | errors: %d\n",
				i+1, requests, elapsed, int(math.Round(rps)), atomic.LoadInt64(&errors))
		} else {
			fmt.Printf("[test-%d]   requests: %d | time(s): %.3f    | req/sec: %d   | too slow: %d | errors: %d\n",
				i+1, requests, elapsed, int(math.Round(rps)), atomic.LoadInt64(&tooSlow), atomic.LoadInt64(&errors))
		}

		client.Close()
		time.Sleep(time.Duration(pauseBetweenRuns) * time.Millisecond)
	}

	if completedExecs > 0 {
		avgRps := int(math.Round(totalRps / float64(completedExecs)))
		fmt.Printf("avg req/sec: %d\n", avgRps)

		if store != "" {
			writeStore(store, parallelism, poolSize, avgRps)
		}
	}
}

func runLatencyTest(url string, warmups, executions int, exercise bool,
	poolSize, timeout, pauseBetweenRuns int, chooser *scriptChooser) {

	fmt.Println("-----------------------LATENCY TEST SELECTED----------------------")

	if exercise {
		fmt.Println("--------------------------INITIALIZATION--------------------------")
		initializeGraph(url, poolSize)
	}

	fmt.Println("---------------------------WARMUP CYCLE---------------------------")
	timeoutSec := float64(timeout) / 1000.0

	for i := 0; i < warmups; i++ {
		client, err := createClient(url, poolSize)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error creating client for warmup: %v\n", err)
			os.Exit(1)
		}

		start := time.Now()
		s := chooser.next()
		rs, err := client.Submit(s)
		if err == nil {
			_, _ = rs.All()
		}
		elapsed := time.Since(start).Seconds()
		fmt.Printf("[warmup-%d]  time: %.6f\n", i+1, elapsed)

		client.Close()

		if elapsed > timeoutSec {
			fmt.Printf("Warmup latency (%.6f s) exceeds timeout (%.3f s), skipping test cycles\n",
				elapsed, timeoutSec)
			return
		}
		time.Sleep(time.Duration(pauseBetweenRuns) * time.Millisecond)
	}

	fmt.Println("----------------------------TEST CYCLE----------------------------")
	var totalLatency float64
	var completedExecs int
	testStart := time.Now()

	for i := 0; i < executions; i++ {
		if time.Since(testStart).Milliseconds() > int64(timeout) {
			fmt.Printf("Timeout reached, skipping remaining test cycles\n")
			break
		}

		client, err := createClient(url, poolSize)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error creating client for test: %v\n", err)
			os.Exit(1)
		}

		start := time.Now()
		s := chooser.next()
		rs, err := client.Submit(s)
		var resultCount int
		if err == nil {
			results, _ := rs.All()
			resultCount = len(results)
		}
		elapsed := time.Since(start).Seconds()
		totalLatency += elapsed
		completedExecs++

		fmt.Printf("[test-%d]  time: %.6f, result count: %d\n", i+1, elapsed, resultCount)

		client.Close()
		time.Sleep(time.Duration(pauseBetweenRuns) * time.Millisecond)
	}

	if completedExecs > 0 {
		avgLatency := totalLatency / float64(completedExecs)
		fmt.Printf("avg latency (sec/req): %.6f\n", avgLatency)
	}
}

func writeStore(storePath string, parallelism, poolSize, rps int) {
	fi, statErr := os.Stat(storePath)
	writeHeader := statErr != nil || fi.Size() == 0

	f, err := os.OpenFile(storePath, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening store file: %v\n", err)
		return
	}
	defer f.Close()

	if writeHeader {
		fmt.Fprintf(f, "parallelism\tpool_size\trequest_per_second\n")
	}
	fmt.Fprintf(f, "%d\t%d\t%d\n", parallelism, poolSize, rps)
}

File: gremlin-python/src/main/python/gremlin_python/driver/util/profiling_application.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
import argparse
import json
import os
import random
import sys
import threading
import time
import traceback
from concurrent.futures import ThreadPoolExecutor, TimeoutError as FuturesTimeoutError, wait

from gremlin_python.driver.client import Client
from gremlin_python.driver.serializer import (
    GraphBinarySerializersV1,
    GraphSONSerializersV2d0,
    GraphSONSerializersV3d0,
)

SERIALIZERS = {
    "graphbinaryv1": GraphBinarySerializersV1,
    "graphsonv2": GraphSONSerializersV2d0,
    "graphsonv3": GraphSONSerializersV3d0,
}

SCRIPTS = [
    "g.V()",
    "g.V(1).out('knows')",
    "g.V(1).out('knows').has('name','josh')",
    'g.V(1).as("a").out("knows").as("b").select("a", "b")',
    'g.V(1).as("a").out("knows").as("b").select("a", "b").by("name")',
    'g.V().hasLabel("person").as("p").map(__.bothE().label().groupCount()).as("r").select("p", "r")',
    'g.V().choose(__.outE().count().is(0L), __.as("a"), __.as("b")).choose(__.select("a"), __.select("a"), __.select("b"))',
    'g.V().group("a").by(T.label).by(outE().values("weight").sum()).cap("a")',
    'g.V().repeat(__.union(__.out("knows").group("a").by("age"), __.out("created").group("b").by("name").by(count())).group("a").by("name")).times(2).cap("a", "b")',
    'g.V().match(as("a").out("knows").as("b"),as("b").out("created").has("name","lop"),as("b").match(as("b").out("created").as("d"),as("d").in("created").as("c")).select("c").as("c")).select("a","b","c")',
]


class ProfilingApplication:

    def __init__(self, execution_name, url, pool_size, max_workers, serializer,
                 requests, executor, script,
                 too_slow_threshold, exercise, suppress_stack_traces):
        self._execution_name = execution_name
        self._url = url
        self._pool_size = pool_size
        self._max_workers = max_workers
        self._serializer = serializer
        self._requests = requests
        self._executor = executor
        self._script = script
        self._too_slow_threshold = too_slow_threshold
        self._exercise = exercise
        self._suppress_stack_traces = suppress_stack_traces
        self._random = random.Random(0)
        # per-execution error count, read by the caller after execute_* runs
        self.errors = 0

    def _create_client(self):
        return Client(self._url, "g",
                      pool_size=self._pool_size,
                      max_workers=self._max_workers,
                      message_serializer=self._serializer)

    def _choose_script(self):
        # Exclude the last (most complex match) script, matching Java behavior
        return SCRIPTS[self._random.randint(0, len(SCRIPTS) - 2)]

    def execute_throughput(self):
        too_slow = [0]
        errors = [0]
        lock = threading.Lock()
        execution_id = "[{}]".format(self._execution_name)
        too_slow_sec = self._too_slow_threshold / 1000.0 if self._too_slow_threshold > 0 else None

        client = self._create_client()
        try:
            def do_request():
                s = self._choose_script() if self._exercise else self._script
                try:
                    result_set = client.submit_async(s).result()
                    result_set.all().result(timeout=too_slow_sec)
                except (FuturesTimeoutError, TimeoutError):
                    with lock:
                        too_slow[0] += 1
                except Exception:
                    with lock:
                        errors[0] += 1
                    if not self._suppress_stack_traces:
                        traceback.print_exc()

            start = time.perf_counter_ns()
            futures = []
            for _ in range(self._requests):
                futures.append(self._executor.submit(do_request))
            wait(futures)
            end = time.perf_counter_ns()

            total_seconds = (end - start) / 1_000_000_000.0
            req_sec = round(self._requests / total_seconds)
            self.errors = errors[0]
            too_slow_display = "N/A" if self._exercise else str(too_slow[0])
            print("{:<10} requests: {} | time(s): {:<14.3f} | req/sec: {:<7} | too slow: {} | errors: {}".format(
                execution_id, self._requests, total_seconds,
                req_sec, too_slow_display, errors[0]), flush=True)
            return req_sec
        except Exception as ex:
            print("Failed Execution: {} - {}".format(self._execution_name, str(ex)))
            self.errors = errors[0] + 1
            return 0
        finally:
            client.close()

    def execute_latency(self, record_errors=False):
        # record_errors=False (warmup default): a request error aborts via
        # RuntimeError, leaving the warmup-gate behavior unchanged.
        # record_errors=True (measured TEST CYCLE): mirror the Go app -- count
        # the error as this execution's error count, still record the measured
        # seconds, and return normally so the caller continues the loop and the
        # single RESULT_JSON line is always emitted. Client creation stays
        # outside the try so a true setup/connection failure remains a hard
        # exit regardless of this flag.
        execution_id = "[{}]".format(self._execution_name)
        client = self._create_client()
        try:
            start = time.perf_counter_ns()
            try:
                result_set = client.submit(self._script)
                size = 0
                for item in result_set:
                    if isinstance(item, list):
                        size += len(item)
                    else:
                        size += 1
            except Exception as ex:
                if not record_errors:
                    if not self._suppress_stack_traces:
                        traceback.print_exc()
                    raise RuntimeError(str(ex)) from ex
                end = time.perf_counter_ns()
                self.errors = 1
                total_seconds = (end - start) / 1_000_000_000.0
                if not self._suppress_stack_traces:
                    traceback.print_exc()
                print("{:<10} time: {:<7}, result count: {}".format(
                    execution_id, "{:.4f}".format(total_seconds), 0), flush=True)
                return total_seconds
            end = time.perf_counter_ns()

            total_seconds = (end - start) / 1_000_000_000.0
            print("{:<10} time: {:<7}, result count: {}".format(
                execution_id, "{:.4f}".format(total_seconds), size), flush=True)
            return total_seconds
        finally:
            client.close()


def _resolve_serializer(name):
    name_lower = name.lower()
    if name_lower not in SERIALIZERS:
        raise ValueError("Unknown serializer: {}. Options: {}".format(
            name, ", ".join(SERIALIZERS.keys())))
    return SERIALIZERS[name_lower]()


def main():
    sys.stdout.reconfigure(line_buffering=True)
    parser = argparse.ArgumentParser(description="Profiling application for gremlin-python driver (3.7)")
    parser.add_argument("--host", default="localhost", help="Server host")
    parser.add_argument("--port", type=int, default=8182, help="Server port")
    parser.add_argument("--parallelism", type=int, default=16, help="Test executor thread pool size")
    parser.add_argument("--test-type", default="throughput", choices=["throughput", "latency"], help="Test mode")
    parser.add_argument("--min-expected-rps", type=int, default=1000, help="Min RPS to continue past warmup")
    parser.add_argument("--timeout", type=int, default=1200000, help="Overall timeout in ms")
    parser.add_argument("--warmups", type=int, default=5, help="Warmup iterations")
    parser.add_argument("--executions", type=int, default=10, help="Test iterations")
    parser.add_argument("--requests", type=int, default=10000, help="Requests per test execution")
    parser.add_argument("--pool-size", type=int, default=8, help="Client connection pool size")
    parser.add_argument("--concurrency", type=int, default=None,
                        help="Canonical concurrency for the cell (provenance passthrough, echoed in RESULT_JSON)")
    parser.add_argument("--max-workers", type=int, default=None, help="Client ThreadPoolExecutor size")
    parser.add_argument("--too-slow-threshold", type=int, default=125, help="Per-request timeout ms")
    parser.add_argument("--serializer", default="graphbinaryv1",
                        choices=list(SERIALIZERS.keys()), help="Serializer")
    parser.add_argument("--pause-between-runs", type=int, default=1000, help="Pause between executions ms")
    parser.add_argument("--suppress-stack-traces", action="store_true", help="Suppress errors")
    parser.add_argument("--exercise", action="store_true", help="Random scripts + Modern graph")
    parser.add_argument("--script", default="1+1", help="Script to execute")
    parser.add_argument("--store", default=None, help="TSV output file path")
    parser.add_argument("--transport", default="ws", choices=["http", "ws"], help="Transport protocol")
    parser.add_argument("--no-exit", action="store_true", help="Don't call sys.exit()")
    args = parser.parse_args()

    executor = ThreadPoolExecutor(max_workers=args.parallelism)
    url = "{}://{}:{}/gremlin".format(args.transport, args.host, args.port)
    serializer_instance = _resolve_serializer(args.serializer)

    # Raw per-execution values collected during the TEST CYCLE only (warmups
    # excluded). Echoed verbatim in the single RESULT_JSON line; no stats here.
    measurements = []
    errors_per_execution = []

    try:
        if args.test_type == "latency":
            print("-----------------------LATENCY TEST SELECTED----------------------")
        else:
            print("---------------------THROUGHPUT TEST SELECTED---------------------")

        if args.exercise:
            print("--------------------------INITIALIZATION--------------------------")
            init_client = Client(url, "g",
                                 pool_size=args.pool_size,
                                 max_workers=args.max_workers,
                                 message_serializer=serializer_instance)
            try:
                init_client.submit("graph.clear()")
                print("Cleared existing 'graph'")
                init_client.submit("TinkerFactory.generateModern(graph)")
                print("Modern graph loaded")
            finally:
                init_client.close()

        if args.test_type == "throughput":
            f = args.store
            if f is not None and (not os.path.exists(f) or os.path.getsize(f) == 0):
                with open(f, "w") as fh:
                    fh.write("parallelism\tpool_size\tmax_workers\trequest_per_second\n")

            meets_rps_expectation = True
            print("---------------------------WARMUP CYCLE---------------------------")
            for ix in range(args.warmups):
                if not meets_rps_expectation:
                    break
                app = ProfilingApplication(
                    "warmup-{}".format(ix + 1), url, args.pool_size,
                    args.max_workers, serializer_instance,
                    1000, executor, args.script, args.too_slow_threshold,
                    args.exercise, args.suppress_stack_traces)
                avg_rps = app.execute_throughput()
                meets_rps_expectation = avg_rps >= args.min_expected_rps
                time.sleep(args.pause_between_runs / 1000.0)

            exceeded_timeout = False
            total_requests_per_second = 0
            completed_executions = 0

            if args.exercise or meets_rps_expectation:
                start = time.perf_counter_ns()
                print("----------------------------TEST CYCLE----------------------------")
                for ix in range(args.executions):
                    if exceeded_timeout:
                        break
                    app = ProfilingApplication(
                        "test-{}".format(ix + 1), url, args.pool_size,
                        args.max_workers, serializer_instance,
                        args.requests, executor, args.script,
                        args.too_slow_threshold, args.exercise,
                        args.suppress_stack_traces)
                    rps = app.execute_throughput()
                    measurements.append(rps)
                    errors_per_execution.append(app.errors)
                    total_requests_per_second += rps
                    completed_executions += 1
                    elapsed_ns = time.perf_counter_ns() - start
                    exceeded_timeout = elapsed_ns > args.timeout * 1_000_000
                    time.sleep(args.pause_between_runs / 1000.0)

            if not meets_rps_expectation or exceeded_timeout or completed_executions == 0:
                avg_rps = 0
            else:
                avg_rps = round(total_requests_per_second / completed_executions)
            print("avg req/sec: {}".format(avg_rps))

            if f is not None:
                with open(f, "a") as fh:
                    fh.write("{}\t{}\t{}\t{}\n".format(
                        args.parallelism, args.pool_size,
                        args.max_workers, avg_rps))

        elif args.test_type == "latency":
            meets_timeout_expectation = True
            print("---------------------------WARMUP CYCLE---------------------------")
            for ix in range(args.warmups):
                if not meets_timeout_expectation:
                    break
                app = ProfilingApplication(
                    "warmup-{}".format(ix + 1), url, args.pool_size,
                    args.max_workers, serializer_instance,
                    1000, executor, args.script, args.too_slow_threshold,
                    args.exercise, args.suppress_stack_traces)
                latency = app.execute_latency()
                meets_timeout_expectation = latency < (args.timeout / 1000.0)
                time.sleep(args.pause_between_runs / 1000.0)

            exceeded_timeout = False
            total_time = 0.0
            completed_executions = 0

            if args.exercise or meets_timeout_expectation:
                start = time.perf_counter_ns()
                print("----------------------------TEST CYCLE----------------------------")
                for ix in range(args.executions):
                    if exceeded_timeout:
                        break
                    app = ProfilingApplication(
                        "test-{}".format(ix + 1), url, args.pool_size,
                        args.max_workers, serializer_instance,
                        args.requests, executor, args.script,
                        args.too_slow_threshold, args.exercise,
                        args.suppress_stack_traces)
                    latency = app.execute_latency(record_errors=True)
                    measurements.append(latency)
                    errors_per_execution.append(app.errors)
                    total_time += latency
                    completed_executions += 1
                    elapsed_ns = time.perf_counter_ns() - start
                    exceeded_timeout = elapsed_ns > args.timeout * 1_000_000
                    time.sleep(args.pause_between_runs / 1000.0)

            if not meets_timeout_expectation or exceeded_timeout or completed_executions == 0:
                avg_latency = 0.0
            else:
                avg_latency = total_time / completed_executions
            print("avg latency (sec/req): {}".format(avg_latency))

        # Single machine-readable result line (see bench/SCHEMA.md). Always
        # printed exactly once; measurements/errors are empty if the test cycle
        # was skipped (warmup gate failed) or cut short (timeout). Warmups are
        # excluded. No statistics are computed here.
        result_payload = {
            "glv": "python",
            "metric": args.test_type,
            "script": args.script,
            "concurrency": args.concurrency,
            "pool": args.pool_size,
            "parallelism": args.parallelism,
            "requests": args.requests,
            "warmups": args.warmups,
            "executions": args.executions,
            "measurements": measurements,
            "errors": errors_per_execution,
        }
        print("RESULT_JSON: " + json.dumps(result_payload), flush=True)

        if not args.no_exit:
            sys.exit(0)
    except Exception:
        traceback.print_exc()
        if not args.no_exit:
            sys.exit(1)
    finally:
        executor.shutdown(wait=True)


if __name__ == "__main__":
    main()
